src/lib/rbtree.zig

```diff
@@ -0,0 +1,973 @@
+        pub fn color_or_black(self: ?*const @This()) Color {
+            return if (self) |self_nonnull| self_nonnull.color else Color.Black;
+        }
+
+        // Get node's sibling, or return null if there is none
+        pub fn sibling(self: *const @This()) ?*@This() {
+            const pos = self.direction();
+            return if (self.par) |parent| parent.desc[1 - pos] else null;
+        }
+    };
+}
+
+/// Red black tree itself
+pub fn Tree(comptime T: type, comptime member_name: []const u8, comptime cfg: Config) type {
+    const NodeType = Node(cfg.features);
+    const enable_iterators_cache = cfg.features.enable_iterators_cache;
+    const enable_kth_queries = cfg.features.enable_kth_queries;
+    // Make sure that comparator type is valid
+    if (!@hasDecl(cfg.comparator, "compare")) {
+        @compileError("Comparator should define \"compare\" function");
+    }
+    const expected_cmp_type = fn (*const cfg.comparator, *const T, *const T) bool;
+    if (!(@TypeOf(cfg.comparator.compare) == expected_cmp_type)) {
+        @compileError("Invalid type of \"compare\" function");
+    }
+    // Make sure that augment callback is valid
+    if (cfg.augment_callback) |augment_callback| {
+        if (!@hasDecl(augment_callback, "augment")) {
+            @compileError("Augment callback should define \"augment\" function");
+        }
+        const expected_augment_callback_type = fn (*const augment_callback, *T) bool;
+        if (!(@TypeOf(augment_callback.augment) == expected_cmp_type)) {
+            @compileError("Invalid type of \"augment\" function");
+        }
+    }
+    return struct {
+        /// Used inside extensions
+        const TreeType = @This();
+        /// root of the tree
+        root: ?*NodeType,
+        /// comparator used for insertions/deletitions
+        comparator: cfg.comparator,
+        /// callback used for augmenting
+        augment_callback: if (cfg.augment_callback) |callback| callback else void,
+        /// size of the tree,
+        size: usize,
+        /// iterators extension data and functions
+        iterators: if (enable_iterators_cache)
+            struct {
+                ends: [2]?*NodeType,
+                /// First element in a tree
+                pub fn first(self: *const @This()) *T {
+                    return node_to_val(self.iterators.ends[0]);
+                }
+
+                /// Last element in a tree
+                pub fn last(self: *const @This()) *T {
+                    return node_to_val(self.iterators.ends[1]);
+                }
+            }
+        else
+            void,
+        /// kth queries extensions
+        kth: if (enable_kth_queries)
+            struct {
+                /// Fix zero sized pointer issue in zig compiler
+                /// Citation: "the issue is: kth is a zero-size type, so the pointer to kth is itself zero-sized, and therefore has no data"
+                /// We need to make it not zero-sized XD
```

Comment:
Maybe leave a link to the upstream issue here so that we can see when we can remove it

Reply:
Issue in ziglang. I am not sure if there is one, but I will search for it.